### DISCUSSION TITLE
3D adjoint optimization

### DISCUSSION THREAD

**Discussion by Paul18811** [NanoComp/meep#2051] [Q&A] [answered]
Hello!

I have done an adjoint optimization of a splitter in 2D and reduced the losses from 3.92 to 3.08 dB for the center wavelength and intermediate design (not dilated or eroded). When 'checking' the result in 3D, the losses are suddenly 5.38 dB (used to be ~ 3.43). This is probably due to an abrupt transition between the waveguides outside of the design area and the optimized design. Some additional optimizations are therefore needed in 3D to account for this well. I however have a question on how to implement this.

If I change the system so that it will perform a 3D calculation, it will also calculate the gradient for every pixel in 3D. The code however shouldn't be able to change pixels in 3D. It should change a pixel in 2D and then apply it for the pixels of that point in the 3th dimension. Is there a way in which I can implement this?

Thank you in advance!

**Comment by smartalecH** [comment, marked as answer]
> The code however shouldn't be able to change pixels in 3D. It should change a pixel in 2D and then apply it for the pixels of that point in the 3th dimension. Is there a way in which I can implement this?

Sure, we do this rather often. If you assign a 2D `MaterialGrid` to a 3D `Block` object, the design parameters are automatically projected onto the third dimension. The proper gradient is automatically computed too.

There are more details in Section 3 of our [recent paper](https://doi.org/10.1364/OE.442074).

**Reply by Paul18811** [reply]
Thank you for your answer!!

This is much more easy than the way in which I implemented it yesterday and it is probably also more accurate.

I first implemented it as doing a 3D simulation with 3D gradients, then averaging the gradients in the 3th dimension to get 2D gradients. These 2D gradients are used to change the geometry of the 2D array. In the next iteration, this  2D array is again translated to 3D and the response/gradients are calculated again.
